### PR TITLE
IDEX-3674: Add git compare with latest repo version functionality

### PR DESCRIPTION
**Pull request by vinokurig**
@skabashnyuk @vparfonov @sleshchenko please review

### CODE REVIEW

File: plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/action/CompareWithLatestRepositoryAction.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import org.eclipse.che.ide.ui.dialogs.ConfirmCallback;
+import org.eclipse.che.ide.ui.dialogs.DialogFactory;
+
+import java.util.Collections;
+
+import static org.eclipse.che.api.git.shared.DiffRequest.DiffType.NAME_STATUS;
+
+/** @author Igor Vinokur */
+@Singleton
+public class CompareWithLatestRepositoryAction extends GitAction {
```
Comment (vzhukovskii) [inline plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/action/CompareWithLatestRepositoryAction.java]: Maybe you should call this class CompareWithHeadAction of smth like this?

File: plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/action/CompareWithLatestRepositoryAction.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    private final NotificationManager     notificationManager;
+    private final GitServiceClient        gitService;
+    private final GitLocalizationConstant locale;
+    private final SelectionAgent          selectionAgent;
+
+    private final String REVISION = "HEAD";
+
+    @Inject
+    public CompareWithLatestRepositoryAction(ComparePresenter presenter,
+                         ChangedListPresenter changedListPresenter,
```
Comment (vzhukovskii) [inline plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/action/CompareWithLatestRepositoryAction.java]: Fix the identations

File: plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/action/CompareWithLatestRepositoryAction.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void actionPerformed(ActionEvent e) {
+        ProjectConfigDto project = appContext.getCurrentProject().getRootProject();
+        String pattern;
+        String path;
+
+        Selection<HasStorablePath> selection = (Selection<HasStorablePath>)selectionAgent.getSelection();
```
Comment (vzhukovskii) [inline plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/action/CompareWithLatestRepositoryAction.java]: Dont't do in this way. Selection agent returns Selection<?>  and you can catch class cast exception/

File: plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/GitExtension.java [hunk 4]
```diff
@@ -48,10 +50,10 @@
 @Singleton
 @Extension(title = "Git", version = "3.0.0")
 public class GitExtension {
-    public static final String GIT_GROUP_MAIN_MENU        = "Git";
-    public static final String REPOSITORY_GROUP_MAIN_MENU = "GitRepositoryGroup";
-    public static final String COMMAND_GROUP_MAIN_MENU    = "GitCommandGroup";
-    public static final String HISTORY_GROUP_MAIN_MENU    = "GitHistoryGroup";
+    public static final String GIT_GROUP_MAIN_MENU           = "Git";
```
Comment (sleshchenko) [inline plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/GitExtension.java]: Why do you add these spaces?

File: plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/action/CompareWithLatestRepositoryAction.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+public class CompareWithLatestRepositoryAction extends GitAction {
+    private final ComparePresenter        comparePresenter;
+    private final ChangedListPresenter    changedListPresenter;
+    private final DialogFactory           dialogFactory;
+    private final NotificationManager     notificationManager;
+    private final GitServiceClient        gitService;
+    private final GitLocalizationConstant locale;
+    private final SelectionAgent          selectionAgent;
+
+    private final String REVISION = "HEAD";
```
Comment (sleshchenko) [inline plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/action/CompareWithLatestRepositoryAction.java]: Why is it not static?

File: plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/compare/ComparePresenter.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+                            @Override
+                            protected void onFailure(Throwable exception) {
+                                notificationManager.showError(exception.getMessage());
+                            }
+                        });
+        }    
+        
+    }
+
+    private void showCompare(final String item, final String oldContent, final String revision) {
```
Comment (sleshchenko) [inline plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/compare/ComparePresenter.java]: private methods should be lower than public

File: plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/compare/CompareView.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+     */
+    void setTitle(String title);
+
+    /** Hide compare panel. */
+    void hide();
+
+    /**
+     * Show compare panel with specified contents.
+     * 
+     * @param oldContent content from specified working tree
```
Comment (sleshchenko) [inline plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/compare/CompareView.java]: Can you format this javadoc in accordance to https://github.com/codenvy/codenvy-dev-resources/wiki/Formatting#javadoc

File: plugin-git/che-plugin-git-ext-git/src/main/resources/org/eclipse/che/ide/ext/git/client/GitLocalizationConstant.properties [hunk 8]
```diff
@@ -143,6 +144,14 @@ view.commit.grid.comment=Comment
 #Create
 view.create.title=Initialize local repository
 
+#Compare
+view.compare.title=with latest repository version
```
Comment (sleshchenko) [inline plugin-git/che-plugin-git-ext-git/src/main/resources/org/eclipse/che/ide/ext/git/client/GitLocalizationConstant.properties]: Strange title of view. Can you check that key has corresponding value

File: plugin-java/che-plugin-java-ext-lang-shared/src/main/java/org/eclipse/che/ide/ext/java/shared/dto/refactoring/GitChanges.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+package org.eclipse.che.ide.ext.java.shared.dto.refactoring;
+
+import org.eclipse.che.dto.shared.DTO;
+
+import java.util.List;
+
+/**
+ * Object represent single change during refactoring.
+ * @author Evgen Vidolob
```
Comment (sleshchenko) [inline plugin-java/che-plugin-java-ext-lang-shared/src/main/java/org/eclipse/che/ide/ext/java/shared/dto/refactoring/GitChanges.java]: Please check author of this class

File: plugin-java/che-plugin-java-ext-lang-shared/src/main/java/org/eclipse/che/ide/ext/java/shared/dto/refactoring/GitChanges.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+ *******************************************************************************/
+
+package org.eclipse.che.ide.ext.java.shared.dto.refactoring;
+
+import org.eclipse.che.dto.shared.DTO;
+
+import java.util.List;
+
+/**
+ * Object represent single change during refactoring.
```
Comment (sleshchenko) [inline plugin-java/che-plugin-java-ext-lang-shared/src/main/java/org/eclipse/che/ide/ext/java/shared/dto/refactoring/GitChanges.java]: Check javadoc

File: plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/compare/ComparePresenter.java (hunk 6 above)
Comment (vzhukovskii) [inline plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/compare/ComparePresenter.java]: actually it doesn't matter :)

Comment (sleshchenko) [inline plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/compare/ComparePresenter.java]: I disagree, but OK :)

Comment (evoevodin) [inline plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/compare/ComparePresenter.java]: Actually it matters!

When i read the class i expect to see its interface first, i want to see its functionality without seen (scrolling) all its private logic, which is not important for me. As far as you may know all public methods must have javadoc, while it is optional for private methods, so if you don't want to make _accordion_ class you have to sort your methods in the right way and let the user get needed information. 

But generally speaking i also think that it is rather organization code standard than language standard. 

My +1 for `public` methods above `private`.

Comment (vzhukovskii) [inline plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/compare/ComparePresenter.java]: Up to you

Comment (evoevodin) [inline plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/compare/ComparePresenter.java]: Well, if you don't mind could you please explain your `actually it doesn't matter :)` ?

Comment (vzhukovskii) [inline plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/compare/ComparePresenter.java]: Sure. If I'll have interface with hundred methods I'll prefer to have internal implementation of this method (private) near public. it's just organization habbit.
For example how it was in 2.x version.
Public api method: https://github.com/codenvy/ide-old/blob/2.22.0/exo-ide-extension-cloudfoundry/src/main/java/org/exoplatform/ide/extension/cloudfoundry/server/Cloudfoundry.java#L291
Internal implementation: https://github.com/codenvy/ide-old/blob/2.22.0/exo-ide-extension-cloudfoundry/src/main/java/org/exoplatform/ide/extension/cloudfoundry/server/Cloudfoundry.java#L298

Comment (azatsarynnyy) [inline plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/compare/ComparePresenter.java]: Agree with Vlad that related things have to be closer together

File: plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/action/CompareWithLatestRepositoryAction.java (hunk 1 above)
Comment (vinokurig) [inline plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/action/CompareWithLatestRepositoryAction.java]: Renamed to CompareWithLatestAction

File: plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/action/CompareWithLatestRepositoryAction.java (hunk 2 above)
Comment (vinokurig) [inline plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/action/CompareWithLatestRepositoryAction.java]: fixed

File: plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/action/CompareWithLatestRepositoryAction.java (hunk 3 above)
Comment (vinokurig) [inline plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/action/CompareWithLatestRepositoryAction.java]: added a checking of instance

File: plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/GitExtension.java (hunk 4 above)
Comment (vinokurig) [inline plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/GitExtension.java]: fixed

File: plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/action/CompareWithLatestRepositoryAction.java (hunk 5 above)
Comment (vinokurig) [inline plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/action/CompareWithLatestRepositoryAction.java]: added static

File: plugin-git/che-plugin-git-ext-git/src/main/resources/org/eclipse/che/ide/ext/git/client/GitLocalizationConstant.properties (hunk 8 above)
Comment (vinokurig) [inline plugin-git/che-plugin-git-ext-git/src/main/resources/org/eclipse/che/ide/ext/git/client/GitLocalizationConstant.properties]: it is a sub-menu of "Compare"

File: plugin-java/che-plugin-java-ext-lang-shared/src/main/java/org/eclipse/che/ide/ext/java/shared/dto/refactoring/GitChanges.java (hunk 9 above)
Comment (vinokurig) [inline plugin-java/che-plugin-java-ext-lang-shared/src/main/java/org/eclipse/che/ide/ext/java/shared/dto/refactoring/GitChanges.java]: removed this unnecessary class

File: plugin-java/che-plugin-java-ext-lang-shared/src/main/java/org/eclipse/che/ide/ext/java/shared/dto/refactoring/GitChanges.java (hunk 10 above)
Comment (vinokurig) [inline plugin-java/che-plugin-java-ext-lang-shared/src/main/java/org/eclipse/che/ide/ext/java/shared/dto/refactoring/GitChanges.java]: removed this unnecessary class

**Pull request closed without merging**